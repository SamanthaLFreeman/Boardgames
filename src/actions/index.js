export const getGames = games => ({
  type: 'GET_GAMES',
  games
})

export const getCategories = categories => ({
  type: 'GET_CATEGORIES',
  categories
})

export const addFavorite = id => ({
  type: 'ADD_FAVORITE',
  id
})

export const removeFavorite = id => ({
  type: 'REMOVE_FAVORITE',
  id
})