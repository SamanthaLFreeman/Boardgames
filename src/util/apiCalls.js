export const getPopularGames = () => {
  return fetch('https://www.boardgameatlas.com/api/search?order_by=popularity&ascending=false&client_id=igbjxf77TF')
    .then(response => {
      if(!response.ok) {
        throw Error('Error fetching popular books')
      }
      return response.json()
    })
}

export const getAllCategories = () => {
  return fetch('https://www.boardgameatlas.com/api/game/categories?client_id=igbjxf77TF')
    .then(response => {
      if(!response.ok) {
        throw Error('Error fetching categories')
      }
      return response.json()
    })
}