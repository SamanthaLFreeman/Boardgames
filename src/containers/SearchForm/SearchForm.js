import React, { Component } from 'react';
import { getGames } from '../../actions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { searchGames, getPopularGames, randomGame } from '../../util/apiCalls';
import { Redirect, Link } from 'react-router-dom';

export class SearchForm extends Component {
  constructor() {
    super();
    this.state = {
      name: "",
      isPushed: false
    }
  }

  handleChange = (e) => {
    this.setState({[e.target.name]: e.target.value})
  }

  handleInputSubmit = (e) => {
    e.preventDefault();
    const { name } = this.state;
    const { getGames, cleanUpGames } = this.props;
    searchGames(name)
      .then(data => cleanUpGames(data.games))
      .then(data => getGames(data))
      .then(() => console.log(this.props.history.push('/favorite')))
      .catch(error => console.log(error))
    this.setState({name: "", isPushed: true})
  }

  handlePopularSubmit = (e) => {
    e.preventDefault();
    const { getGames, cleanUpGames } = this.props;
    getPopularGames()
      .then(data => cleanUpGames(data.games))
      .then(data => getGames(data))
      .catch(error => console.log(error))
    this.setState({ isPushed: true })
  }

  handleRandomSubmit = (e) => {
    e.preventDefault();
    const { getGames, cleanUpGames } = this.props;
    randomGame()
      .then(data => cleanUpGames([data.game]))
      .then(data => getGames(data))
      .catch(error => console.log(error))
    this.setState({ isPushed: true })
  }
  

  render() {
    if(this.state.isPushed) {
      return <Redirect to='/' />
    }
    return (
      <form className="search-form">
        <input 
          className="search search-input"
          type="text" 
          placeholder="Search by name"
          name="name"
          value={this.state.name}
          onChange={this.handleChange} />
        <button 
          className="search"
          onClick={this.handleInputSubmit}>Submit Search</button>
        <button 
          className="search"
          onClick={this.handlePopularSubmit}>Show Popular Games</button>
        <button 
          className="search"
          onClick={this.handleRandomSubmit}>Show Random Game</button>
      </form>
    )
  }
}

export const mapDispatchToProps = dispatch => (
  bindActionCreators({ getGames }, dispatch)
)

export default connect(null, mapDispatchToProps)(SearchForm);