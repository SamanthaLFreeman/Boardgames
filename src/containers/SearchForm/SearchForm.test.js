import React from 'react';
import { shallow } from 'enzyme';
import { SearchForm, mapDispatchToProps } from '../SearchForm/SearchForm';
import { getGames } from '../../actions';
import { getPopularGames, searchGames, randomGame } from '../../util/apiCalls';

jest.mock('../../util/apiCalls');

getPopularGames.mockImplementation(() => {
  return Promise.resolve({
    id: 42,
    name: "Catan",
    year_published: 1999,
    min_players: 2,
    max_players: 6,
    min_playtime: 30,
    max_playtime: 60,
    description_preview: "Gotta get all the sheep",
    image_url: "image",
    primary_publisher: "me",
    categories: [{ id: "1234" }],
    rules_url: "rules link"
  })
});

searchGames.mockImplementation(() => {
  return Promise.resolve({
    id: 42,
    name: "Catan",
    year_published: 1999,
    min_players: 2,
    max_players: 6,
    min_playtime: 30,
    max_playtime: 60,
    description_preview: "Gotta get all the sheep",
    image_url: "image",
    primary_publisher: "me",
    categories: [{ id: "1234" }],
    rules_url: "rules link"
  })
});

randomGame.mockImplementation(() => {
  return Promise.resolve({
    id: 42,
    name: "Catan",
    year_published: 1999,
    min_players: 2,
    max_players: 6,
    min_playtime: 30,
    max_playtime: 60,
    description_preview: "Gotta get all the sheep",
    image_url: "image",
    primary_publisher: "me",
    categories: [{ id: "1234" }],
    rules_url: "rules link"
  })
});

describe('SearchForm', () => {
  let wrapper
  const cleanUpGamesMock = jest.fn();
  const getGamesMock = jest.fn();
  const mockSearchGames = [{
    name: 'Catan',
    id: 42
  }]
  const mockPopularGames = [{
    name: 'Dead of Winter',
    id: 43
  }]
  const mockRandomGame = {
    name: 'Ticket to Ride',
    id: 44
  }

  beforeEach(() => {
    wrapper = shallow(<SearchForm   
      cleanUpGames={cleanUpGamesMock}
      getGames={getGamesMock} />)
  });

  it('should match the snapshot with all of the data passed through it', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when handleChange is called', () => {
    const mockEvent = { target: { name: 'name', value: 'Catan' }};
    const expected = 'Catan';

    wrapper.instance().handleChange(mockEvent);

    expect(wrapper.state('name')).toEqual(expected);
  });

  it('should run handleInputSubmit when the button is clicked', () => {
    wrapper.instance().handleInputSubmit = jest.fn();
    const mockEvent = {preventDefault: jest.fn()};
    wrapper.instance().forceUpdate();
    wrapper.find('button').at(0).simulate('click', mockEvent);

    expect(wrapper.instance().handleInputSubmit).toHaveBeenCalled();
  });

  it('should run handlePopularSubmit when the button is clicked', () => {
    wrapper.instance().handlePopularSubmit = jest.fn();
    const mockEvent = { preventDefault: jest.fn() };
    wrapper.instance().forceUpdate();
    wrapper.find('button').at(1).simulate('click', mockEvent);

    expect(wrapper.instance().handlePopularSubmit).toHaveBeenCalled();
  });

  it('should run handleRandomSubmit when the button is clicked', () => {
    wrapper.instance().handleRandomSubmit = jest.fn();
    const mockEvent = { preventDefault: jest.fn() };
    wrapper.instance().forceUpdate();
    wrapper.find('button').at(2).simulate('click', mockEvent);

    expect(wrapper.instance().handleRandomSubmit).toHaveBeenCalled();
  });

  it('it calls dispatch with the getGames action with the searchedGames data', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = getGames(mockSearchGames);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.getGames(mockSearchGames);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('it calls dispatch with the getGames action with the getPopularGames data', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = getGames(mockPopularGames);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.getGames(mockPopularGames);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('it calls dispatch with the getGames action with the randomGame data', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = getGames(mockRandomGame);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.getGames(mockRandomGame);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('should call searchGames when handleInputSubmit is called', async () => {
    const mockEvent = { preventDefault: jest.fn() }
    wrapper.instance().handleInputSubmit(mockEvent);
    expect(searchGames).toHaveBeenCalled();
  });

  it('should call getPopularGames when handlePopularSubmit is called', async () => {
    const mockEvent = { preventDefault: jest.fn() }
    wrapper.instance().handlePopularSubmit(mockEvent);
    expect(getPopularGames).toHaveBeenCalled();
  });

  it('should call randomGame when handleRandomSubmit is called', async () => {
    const mockEvent = { preventDefault: jest.fn() }
    wrapper.instance().handleRandomSubmit(mockEvent);
    expect(randomGame).toHaveBeenCalled();
  });
})