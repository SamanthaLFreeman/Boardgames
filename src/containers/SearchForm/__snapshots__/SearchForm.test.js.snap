// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot with all of the data passed through it 1`] = `
<form
  className="search-form"
>
  <input
    className="search search-input"
    name="name"
    onChange={[Function]}
    placeholder="Search by name"
    type="text"
    value=""
  />
  <button
    className="search search-btn"
    onClick={[Function]}
  >
    Submit Search
  </button>
  <button
    className="search search-btn"
    onClick={[Function]}
  >
    Show Popular Games
  </button>
  <button
    className="search search-btn"
    onClick={[Function]}
  >
    Show Random Game
  </button>
</form>
`;
